brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/TracerContext.java

```diff
@@ -0,0 +1,100 @@
+package com.github.kristofa.brave.cxf3;
+
+import com.github.kristofa.brave.Brave;
+import com.github.kristofa.brave.ServerSpan;
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+import org.apache.cxf.message.Message;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Used to handle {@link ServerSpan} retrieved from cxf message exchange.
+ */
+public class TracerContext {
+
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+  private final Message message;
+
+  TracerContext(Brave brave, Message message) {
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+    this.message = message;
+  }
+
+  /**
+   * Wraps provided {@link Runnable and ensures same {@link ServerSpan} context as the thread from which the Runnable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Runnable}
+   * @return wrapped {@link Runnable}
+   */
+  public Runnable wrap(Runnable command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        command.run();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  /**
+   * Wraps provided {@link Callable} and ensures same {@link ServerSpan} context as the thread from which the Callable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Callable}
+   * @return wrapped {@link Callable}
+   */
+  public <V> Callable<V> wrap(Callable<V> command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        return command.call();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  private ServerSpan getCurrentServerSpan() {
+    // gets existing server span or reinitializes to ServerSpan#EMPTY using ThreadLocal#initialValue
```

Comment:
PS @reta in brave 4, this could become nice and explicit because Span is always not null (ex it can be a noop)

```java
public void foo(@Suspended AsyncResponse response, @Context brave.Span currentSpan) {
   // use any utility which can act on a span!
}
```

This is why I'm hesitant to adding more things to the api, especially ones that make things that didn't exist yet in Brave 3 explicit. For example, it would be lovely to have an interceptor for today which is wired with a feature and its only role is to attach the brave 3 current span when there's a suspended response. That would mean the users don't have to use any apis which would be almost immediately deprecated!